### PR TITLE
PERF-#5296: Partition parquet file if it has too few row groups

### LINKED ISSUE #5296: Modin doesn't partition DataFrames read from a parquet file if the file itself isn't partitioned
The current implementation of `.read_parquet()` completely relies on the partitioning provided by a parquet file scheme (row_groups, column chunking), this partitioning may not always be good though.

The recommended Apache's configuration for a row_group size considers an optimal row chunk size to be 1GB [[1]](https://parquet.apache.org/docs/file-format/configurations/#row-group-size) which acts terribly for Modin because of the risk to put the whole dataframe into only 1 or 2 partitions, this way of partitioning causes good-parallelized implementations to perform very poorly in comparison with what they could achieve with proper partitioning.

Consider this example where I read a single-row-group parquet file with only 100_000 rows and then apply a simple function to the dataframe. It appears that the dataframe read with the parquet partitioning scheme performs 4x slower than the properly partitioned one. The difference in performance will grow with increasing the dataset.

```
data_shape: (100_000, 2)
parquet num_row_groups: 1
default partitioning: (1, 2)
        .apply() with default partitioning: 13.06s.
proper partitioning: (112, 1)
        .apply() with default partitioning: 3.35s.
```

<details><summary>Reproducer</summary>

```python
import pandas
import modin.pandas as pd
from modin.utils import try_cast_to_pandas
import numpy as np
import tempfile
import pyarrow as pa
from timeit import default_timer as timer

NROWS = 100_000
NCOLS = 2

with tempfile.NamedTemporaryFile() as file:
    pandas.DataFrame({f"col{i}": np.arange(NROWS) for i in range(NCOLS)}).to_parquet(file.name)

    df = pd.read_parquet(file.name)

    parquet = pa.parquet.ParquetFile(file.name)
    print(f"parquet num_row_groups: {parquet.num_row_groups}")
    print(f"default partitioning: {df._query_compiler._modin_frame._partitions.shape}")

    t1 = timer()
    df.apply(lambda row: row + row, axis=1) # no parallelism at all due to poor partitioning
    print(f"\t.apply() with default partitioning: {timer() - t1}")

    df = pd.DataFrame(try_cast_to_pandas(df)) # force repartitioning
    print(f"proper partitioning: {df._query_compiler._modin_frame._partitions.shape}")

    t1 = timer()
    df.apply(lambda row: row + row, axis=1) # decent parallelism
    print(f"\t.apply() with default partitioning: {timer() - t1}")

```

</details>

I wonder if we could to not fully rely on the partitioning provided by a parquet file by using our own partitioning scheme if the provided one is not good enough.

### LINKED ISSUE #5926: Support grouping on a non-column grouper in reshuffling groupby
The reshuffling groupby implementation introduced in ([#put_the_name_of_the_pr](https://github.com/dchigarev/modin/tree/issue_5867)) doesn't support grouping on an arbitrary grouper and raises a warning:
```python
import modin.pandas as pd

df = pd.DataFrame({"a": [1, 2, 3, 4], "b": [3, 4, 5, 6]})
sr = pd.Series([1, 1, 2, 2])

import modin.config as cfg
cfg.ExperimentalGroupbyImpl.put(True)

print(df.groupby(sr).sum())
# UserWarning: Can't use experimental reshuffling groupby implementation because of:
#   Reshuffling groupby is only supported when grouping on a column(s) of the same frame.
#   Falling back to a TreeReduce implementation.
```

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?

The PR corrects unbalanced partitioning after `.read_parquet()`. It basically applies the logic of [`qc.repartition()`](https://github.com/modin-project/modin/blob/14452a8414bdec10e3b5cfa05e98bd26c6e1bafc/modin/core/storage_formats/base/query_compiler.py#L6711) but at the partitions level (for some reason, this implementation works much faster):

image

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5296 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/io/column_stores/parquet_dispatcher.py [hunk 1]
```diff
@@ -646,6 +646,110 @@ def build_index(cls, dataset, partition_ids, index_columns, filters):
...
+        Normalize partitioning according to the default partitioning scheme in Modin.
+
+        The result of 'read_parquet()' is often under partitioned over rows and over partitioned
+        over columns, so this method expands the number of row splits and shrink the number of column splits.
+
+        Parameters
+        ----------
+        remote_parts : np.ndarray
+        row_lengths : list of ints or None
+            Row lengths, if 'None' won't repartition across rows.
```
Comment (YarShev) [inline modin/core/io/column_stores/parquet_dispatcher.py:661]:
```suggestion
            Row lengths, if 'None', won't repartition across rows.
```

File: modin/core/io/column_stores/parquet_dispatcher.py [hunk 2]
```diff
@@ -646,6 +646,110 @@ def build_index(cls, dataset, partition_ids, index_columns, filters):
...
+        remote_parts : np.ndarray
+        row_lengths : list of ints or None
+        column_widths : list of ints
+        """
+        if len(remote_parts) == 0:
+            return remote_parts, row_lengths, column_widths
+
+        from modin.core.storage_formats.pandas.utils import get_length_list
+
+        # The code in this function is actually a duplication of what 'BaseQueryCompiler.repartition()' does,
```
Comment (YarShev) [inline modin/core/io/column_stores/parquet_dispatcher.py:675]: Should we explore this in a separate issue to not have the duplicate code?

File: modin/core/io/column_stores/parquet_dispatcher.py [hunk 3]
```diff
@@ -646,6 +646,110 @@ def build_index(cls, dataset, partition_ids, index_columns, filters):
...
+                if num_splits == 1:
+                    new_parts.append(row_parts)
+                    new_row_lengths.append(part_len)
+                    continue
+
+                offset = len(new_parts)
+                # adding empty row parts according to the number of splits
+                new_parts.extend([[] for _ in range(num_splits)])
+                for part in row_parts:
+                    splited = cls.frame_cls._partition_mgr_cls._column_partitions_class(
```
Comment (YarShev) [inline modin/core/io/column_stores/parquet_dispatcher.py:715]:
```suggestion
                    splitted = cls.frame_cls._partition_mgr_cls._column_partitions_class(
```

File: modin/core/io/column_stores/parquet_dispatcher.py (hunk 2 above)
Comment (dchigarev) [reply modin/core/io/column_stores/parquet_dispatcher.py:675]: created an issue for this #7043

Comment (anmyachev) [reply modin/core/io/column_stores/parquet_dispatcher.py:675]: @dchigarev Looks like you mean `partition_manager.rebalance_partitions`? (not `BaseQueryCompiler.repartition()`)

Comment (dchigarev) [reply modin/core/io/column_stores/parquet_dispatcher.py:675]: no, I meant `qc.repartition()`, I tried using it and it was slower than my implementation

`partition_manager.rebalance_partitions` splits partitions one by one using `partition.mask()`, I also tried this option, and it was also slower than the current implementation

Comment (dchigarev) [reply modin/core/io/column_stores/parquet_dispatcher.py:675]: It is, but I would rather merge it with `rebalance_partitions()` function in order to get a "unified" method to rebalance partitioning #7044

**Review by anmyachev — APPROVED** [review]

**Comment by anmyachev** [comment]
@YarShev any other comments?

**Review by YarShev — APPROVED** [review]

**Pull request merged**